diff --git a/src/afl-fuzz-stats.c b/src/afl-fuzz-stats.c
index 8c13cd6..6a45780 100644
--- a/src/afl-fuzz-stats.c
+++ b/src/afl-fuzz-stats.c
@@ -347,7 +347,7 @@ void write_stats_file(afl_state_t *afl, u32 t_bytes, double bitmap_cvg,
 
   }
 
-#ifndef __HAIKU__
+#if !defined(__HAIKU__) && !defined(__MVS__)
   if (getrusage(RUSAGE_CHILDREN, &rus)) { rus.ru_maxrss = 0; }
 #endif
   u64 runtime_ms = afl->prev_run_time + cur_time - afl->start_time;
@@ -428,7 +428,7 @@ void write_stats_file(afl_state_t *afl, u32 t_bytes, double bitmap_cvg,
           afl->last_crash_time / 1000, afl->last_hang_time / 1000,
           afl->fsrv.total_execs - afl->last_crash_execs, afl->fsrv.exec_tmout,
           afl->slowest_exec_ms,
-#ifndef __HAIKU__
+#if !defined(__HAIKU__) && !defined(__MVS__)
   #ifdef __APPLE__
           (unsigned long int)(rus.ru_maxrss >> 20),
   #else
